hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java

```diff
@@ -222,11 +252,51 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
       throw new RuntimeException("Interrupted during commit", e);
 
     } finally {
-      cleanupMetadataAndUnlock(threw, newMetadataLocation, lockId);
+      cleanupMetadataAndUnlock(commitStatus, newMetadataLocation, lockId);
+    }
+  }
+
+  /**
+   * Attempt to load the table and see if any current or past metadata location matches the one we were attempting
+   * to set. This is used as a last resort when we are dealing with exceptions that may indicate the commit has
+   * failed but are not proof that this is the case. Past locations must also be searched on the chance that a second
+   * committer was able to successfully commit on top of our commit.
+   *
+   * @param newMetadataLocation the path of the new commit file
+   * @param config metadata to use for configuration
+   * @return Commit Status of Success, Failure or Unknown
+   */
+  private CommitStatus checkCommitStatus(String newMetadataLocation, TableMetadata config) {
+    int maxAttempts = PropertyUtil.propertyAsInt(config.properties(), COMMIT_NUM_STATUS_CHECKS,
+        COMMIT_NUM_STATUS_CHECKS_DEFAULT);
+
+    for (int attempt = 1; attempt <= maxAttempts; attempt++) {
```

Comment:
I think this should use `Tasks` instead of a custom loop for retry logic. You can customize the retries and that would keep all of the retry logic in one place.

Reply:
I considered this but I didn't like how I have to set the return value inside of the lambda and can't return from there. I also can't count the number of attempts without another counter, maybe that's a good addition for the task API an "onFailure(item, Exception, Attempt)"

Reply:
I think I'd also really like a "Tasks" API that is a better fit for retry, as is, Tasks really has to propagate through side effects which makes sense for async IO but less for 
output = Try to return something and if you fail return this other thing

Reply:
The tasks API automatically logs retries so you don't need to.

We can look at improving the tasks for retries, but I don't think we need to do that here.